pkg/controller/pxc/users_custom.go

```diff
@@ -0,0 +1,307 @@
+	if cr.Spec.Users == nil && len(cr.Spec.Users) == 0 {
+		return nil
+	}
+
+	if cr.Status.Status != api.AppStateReady {
+		return nil
+	}
+
+	log := logf.FromContext(ctx)
+
+	internalSecrets := corev1.Secret{}
+	err := r.client.Get(context.TODO(),
+		types.NamespacedName{
+			Namespace: cr.Namespace,
+			Name:      internalSecretsPrefix + cr.Name,
+		},
+		&internalSecrets,
+	)
+	if err != nil && !k8serrors.IsNotFound(err) {
+		return errors.Wrap(err, "get internal sys users secret")
+	}
+
+	um, err := getUserManager(cr, &internalSecrets)
+	if err != nil {
+		return err
+	}
+	defer um.Close()
+
+	sysUserNames := sysUserNames()
+
+	for _, user := range cr.Spec.Users {
+		if _, ok := sysUserNames[user.Name]; ok {
+			log.Error(nil, "creating user with reserved user name is forbidden", "user", user.Name)
+			continue
+		}
+
+		if len(user.Grants) == 0 && user.WithGrantOption {
+			log.Error(nil, "withGrantOption is set but no grants are provided", "user", user.Name)
+			continue
+		}
+
+		if user.PasswordSecretRef != nil && user.PasswordSecretRef.Name == "" {
+			log.Error(nil, "passwordSecretRef name is not set", "user", user.Name)
+			continue
+		}
+
+		if user.PasswordSecretRef != nil && user.PasswordSecretRef.Key == "" {
+			user.PasswordSecretRef.Key = "password"
+		}
+
+		if len(user.Hosts) == 0 {
+			user.Hosts = []string{"%"}
+		}
+
+		userSecretName := ""
+		userSecretPassKey := ""
+
+		if user.PasswordSecretRef == nil {
+			userSecretName = fmt.Sprintf("%s-custom-user-secret", cr.Name)
+			userSecretPassKey = user.Name + "-pass"
+
+		} else {
+			userSecretName = user.PasswordSecretRef.Name
+			userSecretPassKey = user.PasswordSecretRef.Key
+		}
+
+		userSecret, err := getUserSecret(ctx, r.client, cr, userSecretName)
+		if err != nil {
+			if k8serrors.IsNotFound(err) {
+				userSecret = &corev1.Secret{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      userSecretName,
+						Namespace: cr.Namespace,
+					},
+				}
+				err := generateUserPass(ctx, r.client, cr, userSecret, userSecretPassKey)
+				if err != nil {
+					return errors.Wrap(err, "failed to generate user password secrets")
+				}
+			} else {
+				log.Error(err, "failed to get user secret", "user", user)
+				continue
+			}
+		}
+
+		annotationKey := fmt.Sprintf("percona.com/%s-%s-hash", cr.Name, user.Name)
+		if userSecret.Annotations == nil {
+			userSecret.Annotations = make(map[string]string)
+		}
+
+		if userPasswordChanged(userSecret, annotationKey, userSecretPassKey) {
+			log.Info("User password changed", "user", user.Name)
+
+			err := um.Exec(ctx, alterUserQuery(&user, string(userSecret.Data[userSecretPassKey])))
+			if err != nil {
+				log.Error(err, "failed to update user", "user", user)
+				continue
+			}
+
+			userSecret.Annotations[annotationKey] = string(sha256Hash(userSecret.Data[userSecretPassKey]))
+			if err := r.client.Update(ctx, userSecret); err != nil {
+				return errors.Wrap(err, "update user secret")
+			}
```

Comment:
What do you think about using `k8s.AnnotateObject` here?

Reply:
I didn't know about this API, I'll use it.

Reply:
[f0fb89d](https://github.com/percona/percona-xtradb-cluster-operator/pull/1835/commits/f0fb89d8542584a63dacae5a5605959428a6b8a0)